plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -310,6 +341,25 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isTopNPushdownEnabled(session) || !jdbcClient.supportsTopNPushDown()) {
+            return Optional.empty();
+        }
+
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+        List<SortItem> sortOrder = sortItems.stream()
+                .map(sortItem -> new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder()))
```

Comment:
- We should pass column handles, not just names, so that the Client can apply logic based on eg type
- verify `assignments.get` returns non-empty and throw nice err when it does